### PR TITLE
S134 Limit nesting depth of control flow statements in functions

### PR DESCRIPTION
**Pull request by p3pijn**
I have made an attempt to implement the rule _Limit nesting depth of control flow statements in functions_ (https://jira.sonarsource.com/browse/RSPEC-134).
I previously suggested this rule in issue #684 in the old repository.

This is my first PR in this repo. Feedback is really appreciated!

### CODE REVIEW

**Comment by p3pijn** [comment]
I figured that I will have to do some additional work to get the regression tests passing.
I'm trying to get the regression test to run on my local machine. I'm running into the following powershell error while running regression-test.bat:

At D:\Git\sonarlint-vs\its\NormalizeSarifReports.ps1:52 char:3
+   Foreach-Object { NormalizeAndSplitSarifReport($_.FullName) }
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ForEach-Object], MethodInvoca
   tionException
    + FullyQualifiedErrorId : ArgumentException,Microsoft.PowerShell.Commands.
   ForEachObjectCommand

ERROR: Error while executing an internal PowerShell script

Does anybody have a clue how to get around this?

**Comment by tvsonar** [comment]
thanks @p3pijn for this contribution. We'll review it.

On your problem with powershell: can you tell us what version are you using? (We are using Win10, and PowerShell 5.0.10586.63)

File: src/SonarLint.CSharp/Rules/FunctionNestingDepth.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => checkNestingDepth<AccessorDeclarationSyntax>(c, a => a.Keyword.GetLocation(), b => b.Body),
+                SyntaxKind.GetAccessorDeclaration,
+                SyntaxKind.SetAccessorDeclaration,
+                SyntaxKind.AddAccessorDeclaration,
+                SyntaxKind.RemoveAccessorDeclaration);
+        }
+
+        private void checkNestingDepth<TSyntax>(SyntaxNodeAnalysisContext c, Func<TSyntax, Location> location, Func<TSyntax, BlockSyntax>  functionBody)
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/FunctionNestingDepth.cs]: Our convention is to upper case all methods, even `private` ones. Could you change the name of this and the following method?

File: src/SonarLint.CSharp/Rules/FunctionNestingDepth.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        private int getNestingDepth(SyntaxNode node, int level)
+        {
+            if (node == null)
+            {
+                return level;
+            }
+
+            var childBlocks = node.DescendantNodes().OfType<BlockSyntax>();
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/FunctionNestingDepth.cs]: `DescendantNodes` get all descendant nodes, so here all top level and nested `BlockSyntax`es are returned. Which means that you could avoid the recursion below.

Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/FunctionNestingDepth.cs]: You'll need to filter a bit more strictly: if you count all `BlockSyntax`es, that will include empty blocks or lambda expressions too, and not just control flow statements as the RSPEC defines.

File: src/Tests/SonarLint.UnitTest/Rules/FunctionNestingDepthTest.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+using SonarLint.Rules.CSharp;
+
+namespace SonarLint.UnitTest.Rules
+{
+    [TestClass]
+    public class FunctionNestingDepthTest
+    {
+        [TestMethod]
+        [TestCategory("Rule")]
+        public void FunctionNestingDepthy()
```
Comment (tvsonar) [inline src/Tests/SonarLint.UnitTest/Rules/FunctionNestingDepthTest.cs]: the method name should be `FunctionNestingDepth`

**Comment by p3pijn** [comment]
Thank you for the feedback @tvsonar.
I will enhance the code according to your comments as soon as I have time.

I was working on Win7 with Powershell 5. I will try to run the regression suite on a Win10 machine.

**Comment by p3pijn** [comment]
@tvsonar I have enhanced this PR according to your feedback. 
Still have to look into the powershell issue though.

**Comment by dbolkensteyn** [comment]
@p3pijn Thanks for the update, but it seems that there is an issue now with this PR : you're trying to add 28 commits, including some that you didn't actually author. The "diff" changes 162 files, making this PR impossible to review.

You'll want to rebase your branch against the latest master one: This will replay your commits on top (i.e. after) all the ones already on the master.

Currently, it looks like you made some commits, then added the ones from the master, and finally made a few further commits. You need all your commits to be made on top of the master branch.

So, you'll probably have to do something like:
```
# Run all this in a new branch to be on the safe side
git checkout -B new_branch
# fetch the latest state of "upstream" (i.e. SonarSource-VisualStudio/sonarlint-roslyn-analyzer - replace "upstream" with the name of your remote for SonarSource-VisualStudio, see git remote -v)
git fetch upstream
# set your "new_branch" branch to the state of SonarSource-VisualStudio/sonarlint-roslyn-analyzer's master branch (again, replace "upstream" if you have to)
git reset --hard upstream/master
# now your branch "new_branch" is at an identical state than SonarSource-VisualStudio/sonarlint-roslyn-analyzer's master branch, push it to your GitHub fork to verify
git push --set-upstream origin new_branch

# go to https://github.com/p3pijn/sonarlint-vs/tree/new_branch to verify manually that everything works as expected

# now, add all your commits on top
git cherry-pick [sha1-of-commit you want to add]

# push the new commits to your fork
git push

# go to https://github.com/p3pijn/sonarlint-vs/tree/new_branch to verify again

# now, you should be able to send a PR from "new_branch" that will look clean
# or, you can also update this one, by forcing your own "master" branch to be at the same state as your "new_branch"

git checkout master
git reset --hard new_branch

# send your updated "master" branch to GitHub, note that you'll need to specify "--force" because your master and the one you have on GitHub will have "divereged" (meaning that you don't share the same "history" of commits). That's fine, "--force" will just overwrite whatever is on GitHub.
git push --force
```

**Comment by p3pijn** [comment]
Thanks a lot for the pointers @Dinesh, I will submit a new clean PR.

**Pull request closed without merging**